Fix broken static linking of apps.

awlauria opened the pull request:
Use pkg-config for both internal and external builds of
libevent, hwloc, and libpmix and store them in
OPAL_WRAPPER_EXTRA_LIBS.

Signed-off-by: Austen Lauria <awlauria@us.ibm.com>

Fixes #8850

awlauria commented:
I tested with internal and external dependencies, both seem to work for static builds and non.

If others can pull this down and independently verify that would be good as well.

awlauria commented:
Some things to iron out, adding the WIP label.

bwbarrett commented on config/opal_check_pc.m4:
```diff
@@ -0,0 +1,40 @@
+         AC_MSG_CHECKING([for pc file for])
+         AC_MSG_RESULT([$1])
+
+	 num_paths=$#
```
I think you have a tab on this one line, because your indenting doesn't line up

awlauria replied:
Got it.

bwbarrett commented on config/opal_check_pc.m4:
```diff
@@ -0,0 +1,40 @@
+
+	 num_paths=$#
+         save_name=$1
+         cmd="pkg-config"
```
This adds an explicit dependency on pkg-config for building, which is new to Open MPI and something I know Jeff was worried about when we talked about it earlier in the week.  We should consider parsing the .pc file directly (as awful as that is, we at least can vaguely contain the pain because we're generating the pc files).

awlauria replied:
I understand not wanting to add another dependency, but trying to parse the file may not be quite good enough - or at worst will be difficult to maintain as things change.

For example, PMIx has a dependency on zlib on our machines. This would require adding `-lz` for static builds, however pmix.pc contains only `requires zlib` as a hint.

Now - for this specific example, it might be easy enough to do that translation, but I don't know if there is a repeatable/defined way for determining the correct linker flag from the library name. It sounds like a maintenance headache, but maybe there is a right way to do it?

Then again, this is exactly what pkg-config is supposed to do, and re-inventing the wheel probably isn't the right approach.

awlauria replied:
As a smaller hammer approach, ompi could only require pkg-config for the `--enable-static` configure line.

bwbarrett replied:
actually, there's no difference between --enable-static or --enable shared here (at least for the external case).  What matters is whether or not PMIX/PRRTE are built with static or not.

bwbarrett replied:
I don't necessarily have objection to using pkg-config directly.  Maybe as a default path, and have an override so users could handle the edge cases or not having pkg-config.  But this is something we should talk about with more input than just you and I :).

jsquyres replied:
I am a bit concerned about requiring `pkg-config`.  It's not present in MacOS, for example.  I don't know if it's available by default in various Linux distros.

Is it really that difficult to parse a `.pc` file?  It's just key=value.  Is it more complicated than `egrep 'THE_RIGHT_KEY\s*=' | cut -f\  -d2`? (disclaimer: typed directly in github and not tested)

awlauria replied:
I don't have a preference either.

@jsquyres we could do it, but do we have a list of all libraries key->value translation that could pop up? See my example on zlib. The pmix.pc file only says, `requires: zlib`, which would require us knowing that an `-lz` is needed for a static build:

```
$. cat pmix.pc
exec_prefix=${prefix}
libdir=${exec_prefix}/lib
includedir=${prefix}/include

Name: pmix
Description: Process Management Interface for Exascale (PMIx)
Version: 5.0.0a1
URL: https://pmix.org/
Requires:  zlib
Libs: -L${libdir} -lpmix 
Cflags: -I${includedir}
```
This is just one example...I don't know of all the translations that would need to be done for every environment. 

Now, if we know all of these, and it's as easy as forming a key -> value table, that would work. I'm not sure what the maintenance would be on that longterm...

You mentioned in the ticket that your hwloc has a dependency on -lpciaccess - that's not something that is required when I build on my machines. It makes me wonder what other things are out there in various environments/setups that will be missed.

awlauria replied:
Well, looking at it again I suppose it could be done:

`Requires: zlib foolib` -> `-lz -lfoo` __should__ work if we don't want the pkg-config dependency.

bwbarrett replied:
I don't think we want to do the requires -> libs change, because we'd lose the recursive lookup behavior.  I kind of like my idea:

1) if pkg-config exists, use it (ie, this CR with all the fixes requested)
2) if pkg-config doesn't exist, assume pmix, prrte, hwloc, libevent, etc. are dynamic linked or have no extra dependent libraries
3) offer a --with-{prrte,pmix,hwloc,libevent}-extra-libs flag so that customers who don't have pkg-config and run into problems with 2 have an exit path.

@jsquyres thoughts?

awlauria replied:
Right, it would be almost impossible for ompi to get the required libs down the line.

That suggestion sounds fine to me. Should it just be a generic `--with-extra-libs` flag? I don't think it really matters that it gets differentiated for purposes of this PR, just as long as they are tacked onto `OMPI_DEPS_LFLAGS`

bwbarrett commented on config/opal_check_pc.m4:
```diff
@@ -0,0 +1,40 @@
+	 num_paths=$#
+         save_name=$1
+         cmd="pkg-config"
+         $1_OMPI_PC_DIR=""
```
You're assuming that $1 is a m4 literal.  That's fine (ie, you don't have to fix that), but assert that it is a literal earlier in the macro so we're not surprised.  Something like:

```
AS_LITERAL_WORD_IF([$1], [], m4_fatal([Non-literal argument $1])])
```

That's completely untested, of course, but I think it's right.

awlauria replied:
Thanks, added.

bwbarrett commented on config/opal_check_pc.m4:
```diff
@@ -0,0 +1,40 @@
+         save_name=$1
+         cmd="pkg-config"
+         $1_OMPI_PC_DIR=""
+         if test $num_paths -gt 1; then
```
a comment about what you're doing here would be appreciated.

awlauria replied:
Updated with comments.

bwbarrett commented on config/opal_check_pc.m4:
```diff
@@ -0,0 +1,40 @@
+dnl
+
+AC_DEFUN([OPAL_CHECK_LFLAGS], [
+         AC_MSG_CHECKING([for pc file for])
```
ew; this is going to be `checking for pc file for... HWLOC`, which is ugly.  Maybe just an `AC_MSG_NOTICE([Parsing pc file for $1])`?

bwbarrett commented on config/opal_check_pc.m4:
```diff
@@ -0,0 +1,40 @@
+         fi
+
+         cmd="${cmd} --libs-only-l $save_name"
+         AC_MSG_CHECKING([pkg-config command for $save_name])
```
same as above; I'd use `AC_MSG_NOTICE` here, and I don't think we need the command in the stdout output stream, but should log it in `config.log`.  Have a look at the `OPAL_LOG_COMMAND` macro.

awlauria replied:
Resolved, that is cleaner.

bwbarrett commented on config/opal_check_pc.m4:
```diff
@@ -0,0 +1,40 @@
+         AC_MSG_CHECKING([pkg-config command for $save_name])
+         AC_MSG_RESULT([$cmd])
+
+         TMP=$($cmd)
```
I wouldn't name a variable `TMP`, as some systems set `$TMP` as location of the temp directory.  Also, you should use OPAL_VAR_PUSH and OPAL_VAR_POP to clean up the environment for such globally named variables.

awlauria replied:
I think I got them all. Thanks

bwbarrett commented on config/opal_config_hwloc.m4:
```diff
@@ -70,6 +70,14 @@ AC_DEFUN([OPAL_CONFIG_HWLOC], [
     AS_IF([test "$external_hwloc_happy" = "0" -a "$internal_hwloc_happy" = "0"],
           [AC_MSG_ERROR([Could not find viable hwloc build.])])
 
+    AS_IF([test "$external_hwloc_happy" = "1"],
```
Please use `test "$opal_hwloc_mode" = "external"` here, as that's the proper indicator for an internal vs. external build.

bwbarrett commented on config/opal_check_pc.m4:
```diff
@@ -0,0 +1,40 @@
+         [
+           AC_MSG_CHECKING([pkg-config result for $save_name])
+           AC_MSG_RESULT([$TMP])])
+           OMPI_DEPS_LFLAGS="$OMPI_DEPS_LFLAGS $TMP"
```
Why are we threading yet another variable, rather than jsut adding these to the wrapper flags?  And you should call `OPAL_WRAPPER_FLAGS_ADD([LIBS], [$TMP])` instead of modifying the environment directly.

bwbarrett commented on config/opal_config_libevent.m4:
```diff
@@ -70,6 +70,23 @@ AC_DEFUN([OPAL_CONFIG_LIBEVENT], [
           [AC_MSG_ERROR([Could not find viable libevent build.])])
 
+    AS_IF([test "$external_libevent_happy" = "1"],
+        [AS_IF([test -n "$with_libevent"],
```
why is this conditional block different than hwloc?

awlauria replied:
ompi requires libevent_core and libevent_pthreads, as a pmix dependency I believe. From what I can tell just checking libevent.pc only returns -levent.

bwbarrett replied:
can you add a comment to that effect (that we need -llibevent_core (the new hotness), but the .pc file only returns -levent (the old brokenness), so the next person doesn't have to figure that out?

awlauria commented:
@bwbarrett thanks for the comments, I think I addressed them all. I also added the --with-extra-libs option to configure.

ibm-ompi commented:
The IBM CI (PGI) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/ibm-ompi/88b480fb98dfd2bdac939f023e1f8fe6

ibm-ompi commented:
The IBM CI (GNU/Scale) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/ibm-ompi/88b480fb98dfd2bdac939f023e1f8fe6

ibm-ompi commented:
The IBM CI (XL) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/ibm-ompi/88b480fb98dfd2bdac939f023e1f8fe6

jjhursey commented:
bot:ibm:retest

awlauria commented on config/opal_check_pc.m4:
```diff
@@ -0,0 +1,109 @@
+          opal_pkg_config_result=$($opal_pkg_config_cmd)
+          AS_IF([test -z "$opal_pkg_config_result"],
+                [
+                  AC_MSG_WARN([Could not find viable $opal_pkg_config_name.pc])
```
@bwbarrett for some reason libevent_core/libevent_pthreads are not located in the default location on aws ci for libevent external builds.

Not sure if they don't exist, or just don't exist where pkg-config looks for them. To get it to pass I changed this from a hard error to a warning.

bwbarrett replied:
I'm pretty sure the external libevent is installed in a user directory (rather than a system directory), which is why pkg-config would be wonky.  This probably isn't an unusual config, so I think your change is the right one.  I'll think about the right way to add some tests for this  (as well as external PMIx/PRRTE) in the coming weeks.

bwbarrett commented on config/opal_check_pc.m4:
```diff
@@ -0,0 +1,109 @@
+dnl
+AC_DEFUN([OPAL_CHECK_LFLAGS], [
+
+  OPAL_CHECK_PKG_CONFIG
```
Any reason this isn't an AC_REQUIRE?

bwbarrett commented on config/opal_check_pc.m4:
```diff
@@ -0,0 +1,109 @@
+                  AC_MSG_WARN([Could not find viable $opal_pkg_config_name.pc])
+                ],
+                [
+                  # Strip -levent as we don't want it. We only want -levent_core/-levent_pthreads.
```
I think this should happen in the calling layer, so all the libevent logic is in one place.  That might mean that the OPLA_FLAGS_APPEND_UNIQ has to be one layer up, but I think that's a better solution.

awlauria replied:
I agree, I moved it up a layer to strip out the -levent.

bwbarrett reviewed (changes requested):
Other than the libevent handling, looks good to go to me.

awlauria commented:
bot:aws:retest

java failures.

gpaulsen reviewed (commented):
I worry that we're blurring the line between ompi and opal too much here, but I'm curious of other's thoughts.

awlauria commented on config/opal_check_pc.m4:
```diff
@@ -0,0 +1,105 @@
+
+dnl OPAL_CHECK_LFLAGS
+dnl Get the -l flags using pkg-config on the passed in .pc file, and
+dnl append them to OMPI_WRAPPER_EXTRA_LIBS.
```
@jsquyres

bwbarrett replied:
I think opal_wrapper_extra_libs is probably marginally more correct, since many of the symbols are used in OPAL.  Point (2) isn't really valid.  In either case, the libraries aren't linked into the libmpi.a file, they're dependencies the user must specify.  And since libmpi.a depends on libopen-pal.so, any transitive inclusion into ABI you can come up with for static libraries means the impact on ABI is the same.

awlauria commented:
@jsquyres @gpaulsen this still needs a review.

awlauria commented:
Ping - this is a v5.0.x item. Please review so this can get in.

bwbarrett commented on config/opal_check_pc.m4:
```diff
@@ -0,0 +1,105 @@
+dnl                      PKG_CONFIG_PATH to tell pkg-config where to locate the .pc
+dnl                      file from argument 1, and its dependencies.
+dnl
+AC_DEFUN([OPAL_CHECK_LFLAGS], [
```
Can we come up with a better name than this?  Something like OPAL_GET_LFLAGS_FROM_PC?

bwbarrett reviewed (approved):
Overall, looks good.  A couple of minor requests included.

awlauria commented:
Rebased with no changes on top of master.

bwbarrett commented:
@awlauria I can't figure out why the MLNX CI failed, it does not look like anything related to this patch and looks like it has something to do with stdout, maybe, but I really can't figure out what they're testing right then.

@janjust can you bug someone to take a look and help diagnose?

awlauria commented:
Thanks - yeah checking myself it looks like an mpirun failed.

I'm still planning to update the PR to address your comments, I was hoping to do that today but got distracted. Probably will update monday and verify it still works.

awlauria commented:
I addressed comments in a separate commit. Biggest change being that I changed these to be appended to OPAL_WRAPPER_EXTRA_LIBS instead of OMPI.

Also rebased against master again.

If this looks good I'll squash and merge.

awlauria commented:
Forgot to sign-off.
